github: Add GitHub action to greet first time contributors

kartben opened the pull request:
Adds the configuration file for a GitHub action that will greet first-time contributor the first time they:

- open an issue
- open a PR
- get a PR of theirs merged

stephanosio commented on .github/config.yml:
```diff
@@ -0,0 +1,9 @@
+firstPRMergeComment: |
+  🎉🥳 Congratulations on getting your very first Zephyr pull request merged! This is a fantastic achievement, and we're thrilled to have you as part of our community!
+
+  To celebrate this milestone and showcase your contribution, we'd love to award you the Zephyr Technical Contributor badge. If you're interested, please claim your badge by filling out this form: (Claim Your Zephyr Badge)[https://forms.gle/oCw9iAPLhUsHTapc8].
```
Incorrect link format

```suggestion
  To celebrate this milestone and showcase your contribution, we'd love to award you the Zephyr Technical Contributor badge. If you're interested, please claim your badge by filling out this form: [Claim Your Zephyr Badge](https://forms.gle/oCw9iAPLhUsHTapc8).
```

stephanosio reviewed (changes requested):
No objection to this in general, but would like a bit more community input before we merge this.

kartben commented:
Thanks for the swift review @stephanosio! And +1 to gather community input (I actually considered opening an RFC but that felt overkill?)
FWIW @carlescufi just suggested that this might be better off implemented as a Github action, e.g. using https://github.com/marketplace/actions/first-interaction

stephanosio commented:
> FWIW @carlescufi just suggested that this might be better off implemented as a Github action, e.g. using https://github.com/marketplace/actions/first-interaction

AFAICS, that action adds a comment when a user opens their first issue/PR, not when the first PR is merged.

mbolivar-nordic commented:
>  As our Continuous Integration jobs run checks on your PR to ensure it's compliant and doesn't cause any issues

@kartben misleading: CI workflows for first-time contributors must be manually approved on every push

kartben commented:
> > As our Continuous Integration jobs run checks on your PR to ensure it's compliant and doesn't cause any issues
> 
> @kartben misleading: CI workflows for first-time contributors must be manually approved on every push

+1, I also caught that, thanks! Will update the wording.

kartben commented:
> > As our Continuous Integration jobs run checks on your PR to ensure it's compliant and doesn't cause any issues
> 
> @kartben misleading: CI workflows for first-time contributors must be manually approved on every push

Actually, I just remembered thinking twice while writing it -- it is in fact the correct wording :) As this is implemented as a Github action, the bot will in fact comment on the PR only when someone will have manually approved CI. If you think that can help people understand better how we work the message could actually be reworded something like "A project maintainer just triggered our CI pipeline to run on your contribution, ...."

kartben commented:
Unfortunately, I haven't heard back from the maintainers of the actions/first-interaction (see actions/first-interaction#226).

Can someone please help me get this fork https://github.com/kartben/first-interaction/tree/v1.1.2-zephyr hosted under the zephyrproject-rtos organization? 
@carlescufi

kartben committed:
github: add GitHub action to greet first time contributors

Adds the configuration file for a GitHub action that will greet first-time
contributor the first time they:
- open an issue
- open a PR
- get a PR of theirs merged

Fixes #56092.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

mbolivar-nordic commented:
> As this is implemented as a Github action, the bot will in fact comment on the PR only when someone will have manually approved CI.

Ah, OK! That makes sense.

> If you think that can help people understand better how we work the message could actually be reworded something like "A project maintainer just triggered our CI pipeline to run on your contribution, ...."

I think that would be an improvement.

kartben commented:
> I think that would be an improvement.

+1. I've fixed the wording.
I am ready to move the PR out of draft state -- would be great if someone can assist me with getting a new repo created into the org for the GH action. Ideally it's created as a actual fork of https://github.com/actions/first-interaction so that it properly shows up in Github as being a fork.

mbolivar-nordic commented:
> would be great if someone can assist me with getting a new repo created into the org for the GH action.

Take the PR out of draft and add TSC label -- if nobody gets to it by the next TSC meeting, it can be handled then

kartben commented:
> > would be great if someone can assist me with getting a new repo created into the org for the GH action.
> 
> Take the PR out of draft and add TSC label -- if nobody gets to it by the next TSC meeting, it can be handled then

Done, thank you!

dleach02 commented on .github/workflows/greet_first_time_contributor.yml:
```diff
@@ -0,0 +1,50 @@
+            first pull request to the Zephyr project!
+
+
+            A project maintainer just triggered our CI pipeline to run it against your your PR and
```
"your your"

kartben replied:
thanks!

kartben commented:
@stephanosio We briefly discussed this in today's TSC call. Could you please create a fork of https://github.com/actions/first-interaction under the Zephyr org and make me a contributor so that I can push my fork there and then update this PR accordingly? Thank you!

stephanosio commented:
> Could you please create a fork of https://github.com/actions/first-interaction under the Zephyr org and make me a contributor so that I can push my fork there and then update this PR accordingly?

@kartben Done. https://github.com/zephyrproject-rtos/action-first-interaction

kartben commented:
> > Could you please create a fork of https://github.com/actions/first-interaction under the Zephyr org and make me a contributor so that I can push my fork there and then update this PR accordingly?
> 
> @kartben Done. https://github.com/zephyrproject-rtos/action-first-interaction

Thanks! I just removed the DNM label as this should now be ready to be fully reviewed and merged.

kartben commented:
> No objection to this in general, but would like a bit more community input before we merge this.

@stephanosio please have another look when you get a chance. Thanks!

nordicjm commented:
I guess my question to this is pretty brief: why?

stephanosio commented on .github/workflows/greet_first_time_contributor.yml:
```diff
@@ -0,0 +1,50 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: zephyrproject-rtos/action-first-interaction@v1.1.2-zephyr
```
This should point to a tag (e.g. `v1.1.2-1`) so that arbitrary pushes to the branch in the future does not break this workflow.

kartben replied:
Great catch, thanks. I just renamed the branch and created a `v1.1.2-zephyr` tag.

kartben replied:
fixed with latest update

kartben commented:
> I guess my question to this is pretty brief: why?

Mmmh I guess I could reply briefly too... why _not_? The idea is simply to encourage first time contributors (which we get on a daily basis) to get more involved and to really make the point that we do value their contributions. While it certainly won't be the silver bullet that will get people to automagically consider becoming long-time contributors, my hope is that it can contribute to getting some of them to come back.

FWIW, out of the 500+ unique individuals who contributed to Zephyr over the last 6 months, roughly 50% contributed 2 commits or less (in fact, 33% contributed only 1). Even if we capture a small portion of these folks, I'll consider this to be a win, but YMMV.

stephanosio reviewed (changes requested):
I am very concerned about the `isFirstPull` function implementation:
https://github.com/zephyrproject-rtos/action-first-interaction/commit/5d55935c19f4b6f45eaa113304c50487f53353fb#diff-7934bf411fea192ad8cd69e0a12911648a2842cb0f2409a8fb67b41b7069d757R153

```typescript
function isFirstPull(client, sender, curPullNumber, closed, page = 1) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
...
        return yield isFirstPull(client, sender, curPullNumber, closed, page + 1);
    });
}
```

Here you are effectively querying every pull request that has ever been created in the repository whenever a pull request is opened or closed, with each query API request returning 100 of them.

We currently have a total of 38670 pull requests in the Zephyr main repository, and **that means 387 API requests per every PR open/close!**

This is simply not going to scale and we will hit the API request limit very quickly.

nordicjm commented:
> FWIW, out of the 500+ unique individuals who contributed to Zephyr over the last 6 months, roughly 50% contributed 2 commits or less (in fact, 33% contributed only 1). Even if we capture a small portion of these folks, I'll consider this to be a win, but YMMV.

I've seen numerous PRs that are a first time contribution for someone fixing a typo or fixing a very minor issue with the documentation, it's a wrong assumption to assume that every contributor is going to submit multiple PRs to the project, maybe they're correcting an issue they saw and other than that they're entirely happy with how the code is. From my perspective I don't see the value this adds, if someone has something to contribute, they can contribute it, if not, then there's nothing to submit.

carlescufi commented:
> > As our Continuous Integration jobs run checks on your PR to ensure it's compliant and doesn't cause any issues
> 
> @kartben misleading: CI workflows for first-time contributors must be manually approved on every push

Not all of them though! Just a subset, which is confusing.

carlescufi commented:
@kartben I have no objections to this PR, thanks for the effort in forking the original action and modifying it to add the functionality we need. The comments from @stephanosio however need to be investigated. I wonder if there is an easier way to find out? For example using https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#get-a-repository to get the contributor list and then check if the user is on that list?

kartben commented:
> We currently have a total of 38670 pull requests in the Zephyr main repository, and **that means 387 API requests per every PR open/close!**
> 
> This is simply not going to scale and we will hit the API request limit very quickly.

I was just starting to think about it and yes, you're probably right :) I got misled by the way the original action was doing things, and the fact that it said that there's no API to filter PRs by creator (see https://github.com/actions/first-interaction/blob/main/src/main.ts#L126) so didn't look/think further. 

I will look at @carlescufi's suggestion but it also looks like I can probably get all I need with the /search API (`is:pr repo:zephyrproject-rtos/zephyr author:janedoe`, for which the rate limit is 30 requests per minute which should be plenty.

stephanosio commented:
> but it also looks like I can probably get all I need with the /search API (`is:pr repo:zephyrproject-rtos/zephyr author:janedoe`, for which the rate limit is 30 requests per minute which should be plenty.

Agreed. The search-based solution sounds quite promising.

The contributor list approach has the following issue;

> This endpoint may return information that is a few hours old because the GitHub REST API caches contributor data to improve performance.

dkalowsk commented:
> > FWIW, out of the 500+ unique individuals who contributed to Zephyr over the last 6 months, roughly 50% contributed 2 commits or less (in fact, 33% contributed only 1). Even if we capture a small portion of these folks, I'll consider this to be a win, but YMMV.
> 
> I've seen numerous PRs that are a first time contribution for someone fixing a typo or fixing a very minor issue with the documentation - and that's the only PR they ever do, it's a wrong assumption to assume that every contributor is going to submit multiple PRs to the project, maybe they're correcting an issue they saw and other than that they're entirely happy with how the code is. From my perspective I don't see the value this adds, if someone has something to contribute, they can contribute it, if not, then there's nothing to submit.

I tend to fall in the same camp here.  I find such automated messages to be annoying and generally a turn off to contributing further.  I also don't see any harm in trying something new to encourage contributions.  If it works, great.  If it doesn't it can be iterated to a better solution or disabled later.  Most of the regular contributors will never even see the message.

kartben commented:
> Agreed. The search-based solution sounds quite promising.

And so much easier too! Action has been updated in this PR: https://github.com/zephyrproject-rtos/action-first-interaction/pull/1

stephanosio commented on .github/workflows/greet_first_time_contributor.yml:
```diff
@@ -0,0 +1,50 @@
+
+jobs:
+  check_for_first_interaction:
+    runs-on: ubuntu-latest
```
Please use `ubuntu-22.04` here as do the rest of the workflows.

We do not want any surprises when GH decided to change the `latest` version in the future.

kartben commented:
> A side effect of this is you can now no longer tell the status of issues submitted, given the following list of queries, is someone dealing with them/asking questions or is it a bot posting a comment?
> 
> ![Screenshot_20230426_092734](https://user-images.githubusercontent.com/40387179/234517588-0994aa4b-2c4b-49ea-8ab6-69e6b842f4d6.png)

I get your point but I'd argue that only using "has 0 comment" as the criteria to determine if someone has engaged is probably not the best approach in the first place as it can miss instances where the original author made several comments // see e.g the following issues, by only looking at the first two pages of opened issues 
https://github.com/zephyrproject-rtos/zephyr/issues/56614
https://github.com/zephyrproject-rtos/zephyr/issues/56670
https://github.com/zephyrproject-rtos/zephyr/issues/56697
https://github.com/zephyrproject-rtos/zephyr/issues/57037
